src/core/tactics.ml

```diff
@@ -56,9 +56,12 @@ let handle_tactic : sig_state -> Proof.t -> p_tactic -> Proof.t =
     set_meta m (Bindlib.unbox (Bindlib.bind_mvar (Env.vars env) (lift t)));
     (* New subgoals and focus. *)
     let metas = Basics.get_metas true t in
-    let new_goals = List.map Proof.Goal.of_meta metas in
+    let add_goal m = List.insert Proof.Goal.compare (Proof.Goal.of_meta m) in
+    let new_goals = MetaSet.fold add_goal metas [] in
+    (* New goals must appear first. *)
     Proof.({ps with proof_goals = new_goals @ gs})
   in
+
```

Comment:
1) Meta keys increase. So metas are ordered in the order of generation. Indeed, using sets didn't break any proof script.
2) Your code snippet is not well typed-checked: a goal is not just a meta.

Reply:
1. Order of generation is probably not the order we want in presence of instantiation. I don't have any precise example in mind, but the fact that this does not break any proof script is not so surprising since we have basically none.
2. Some `List.map` should do the trick.